File: 2020/14/docking.c
#include <stdlib.h>
#include <stdio.h>
#include <inttypes.h>
#include <string.h>

#define u64 uint64_t
#define LSIZE (4 + 3 + 36 + 2)

void toMask(u64 mask[2], const char *str) {
    mask[0] = UINT64_MAX;
    mask[1] = 0;
    for (int i = 0; i < 36; i++) {
        switch (str[i]) {
            case 'X':
                break;
            case '0':
                mask[0] &= ~((u64)1 << (35 - i));
                break;
            case '1':
                mask[1] |= (u64)1 << (35 - i);
                break;
        }
    }
}

int toMem(const u64 mask[2], u64 **mem, u64 *mem_size, const char *str) {
    char *bracket;
    u64 place = strtoumax(str + 4, &bracket, 10);
    u64 value = strtoumax(bracket + 4, NULL, 10);
    value &= mask[0];
    value |= mask[1];

    u64 new_size = place + 1;
    if (new_size >= *mem_size) {
        *mem = realloc(*mem, new_size * sizeof(u64));
        if (!*mem) return 0;
        memset(*mem + *mem_size, 0, (new_size - *mem_size) * sizeof(u64));
        *mem_size = new_size;
    }

    (*mem)[place] = value;

    return 1;
}

int main() {
    u64 mem_size = 0;
    u64 *mem = malloc(mem_size);
    if (!mem) {
        return EXIT_FAILURE;
    }

    u64 mask[2];
    char line[LSIZE];
    while (fgets(line, sizeof line, stdin)) {
        if (line[1] == 'a') {
            toMask(mask, line + 7);
        } else {
            if (!toMem(mask, &mem, &mem_size, line)) {
                return EXIT_FAILURE;
            }
        }
    }

    u64 sum = 0;
    for (int i = 0; i < mem_size; i++) {
        sum += mem[i];
    }

    printf("%" PRIu64 "\n", sum);
}

File: 2020/14/docking.rs
use std::{
    collections::HashMap,
    io::{self, BufRead},
};

type Mem = HashMap<u64, u64>;

fn main() {
    let mut mem1: Mem = HashMap::new();
    let mut mem2: Mem = HashMap::new();
    let mut mask: (u64, u64) = (0, 0);
    let mut masks: Vec<(u64, u64)> = Vec::new();

    for line in io::stdin().lock().lines() {
        let str_line = line.unwrap();
        if str_line.as_bytes()[1] == b'a' {
            mask = to_mask(&str_line[7..43]);
            masks = to_masks(&str_line[7..43]);
        } else {
            let bracket = str_line.find(']').unwrap();
            let place: u64 = str_line[4 .. bracket].parse().unwrap();
            let value: u64 = str_line[bracket + 4 ..].parse().unwrap();

            mem1.insert(place, value & mask.0 | mask.1);
            for m in &masks {
                mem2.insert(place & m.0 | m.1, value);
            }
        }
    }

    println!("{}\n{}", mem1.values().sum::<u64>(),
                       mem2.values().sum::<u64>());
}

fn to_mask(s: &str) -> (u64, u64) {
    let mut mask = (u64::MAX, 0);
    for (i, c) in s.as_bytes().iter().enumerate() {
        match c {
            b'0' => mask.0 &= !(1 << (35 - i)),
            b'1' => mask.1 |= 1 << (35 - i),
            b'X' => {},
            _ => panic!()
        }
    }
    return mask;
}

fn to_masks(s: &str) -> Vec<(u64, u64)> {
    let mut masks = vec![(u64::MAX, 0)];
    for (i, c) in s.as_bytes().iter().enumerate() {
        match c {
            b'0' => {}
            b'1' => for mask in &mut masks {
                mask.1 |= 1 << (35 - i);
            }
            b'X' => {
                let mut m = Vec::new();
                for mask in &mut masks {
                    mask.0 &= !(1 << (35 - i));
                    m.push((mask.0, mask.1 | (1 << (35 - i))));
                }
                masks.append(&mut m);
            }
            _ => panic!()
        }
    }
    return masks;
}
